tests/test_deferredlogger.cpp

```diff
@@ -0,0 +1,94 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+
+#define BOOST_TEST_MODULE TestDeferredLogger
+
+#include <boost/test/unit_test.hpp>
+
+#include "opm/simulators/DeferredLogger.hpp"
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/OpmLog/LogBackend.hpp>
+#include <opm/common/OpmLog/CounterLog.hpp>
+#include <opm/common/OpmLog/TimerLog.hpp>
+#include <opm/common/OpmLog/StreamLog.hpp>
+#include <opm/common/OpmLog/LogUtil.hpp>
+
+using namespace Opm;
+
+void initLogger(std::ostringstream& log_stream) {
+    OpmLog::removeAllBackends();
+    std::shared_ptr<CounterLog> counter = std::make_shared<CounterLog>();
+    std::shared_ptr<StreamLog> streamLog = std::make_shared<StreamLog>( log_stream , Log::DefaultMessageTypes);
+
+    OpmLog::addBackend("COUNTER" , counter);
+    OpmLog::addBackend("STREAM" , streamLog);
+    BOOST_CHECK_EQUAL( true , OpmLog::hasBackend("COUNTER"));
+    BOOST_CHECK_EQUAL( true , OpmLog::hasBackend("STREAM"));
+
+    streamLog->setMessageFormatter(std::make_shared<SimpleMessageFormatter>(true, false));
+    streamLog->setMessageLimiter(std::make_shared<MessageLimiter>(2));
+}
+
+
+BOOST_AUTO_TEST_CASE(deferredlogger)
+{
+
+    const std::string expected = Log::prefixMessage(Log::MessageType::Info, "info 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Warning, "warning 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Error, "error 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Error, "error 2") + "\n"
+        + Log::prefixMessage(Log::MessageType::Problem, "problem 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Bug, "bug 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Debug, "debug 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Note, "note 1") + "\n"
+        + Log::prefixMessage(Log::MessageType::Note, "note 2") + "\n"
+        + Log::prefixMessage(Log::MessageType::Note, "note 3") + "\n";
+
+    std::ostringstream log_stream;
+    initLogger(log_stream);
+    auto deferredlogger = Opm::DeferredLogger();
+    deferredlogger.info("info 1");
+    deferredlogger.warning("warning 1");
+    deferredlogger.error("error 1");
+    deferredlogger.error("error 2");
+    deferredlogger.problem("problem 1");
+    deferredlogger.bug("bug 1");
+    deferredlogger.debug("debug 1");
+    deferredlogger.note("note 1");
+    deferredlogger.note("tagme", "note 2");
+    deferredlogger.note("tagme", "note 3");
```

Comment:
To verify that the tagged messages are forwarded correctly, add another note with the same tag, due to the max limit of 2 it should not be present in the output, instead you should have something about the limit begin reached.